pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java

```diff
@@ -148,4 +156,29 @@ public Consumer select(int hash) {
     Map<Integer, List<Consumer>> getRangeConsumer() {
         return Collections.unmodifiableMap(hashRing);
     }
+
+    @Override
+    public StickyKeyConsumerPredicate generateSpecialPredicate(final Consumer consumer){
+        NavigableMap<Integer, List<String>> cpHashRing = new ConcurrentSkipListMap<>();
+        for (Map.Entry<Integer, List<Consumer>> entry: hashRing.entrySet()) {
+            if (CollectionUtils.isEmpty(entry.getValue())) {
+                continue;
+            }
+            cpHashRing.put(entry.getKey(), entry.getValue()
+                    .stream()
+                    .map(v -> v == consumer ? StickyKeyConsumerPredicate.SPECIAL_CONSUMER_MARK
+                            : StickyKeyConsumerPredicate.OTHER_CONSUMER_MARK)
+                    .collect(Collectors.toList())
+            );
+        }
+        return new StickyKeyConsumerPredicate.Predicate4ConsistentHashingStickyKeyConsumerSelector(cpHashRing);
+    }
+
+    private void notifyActiveConsumerChanged() {
+        // TODO add configuration for this events in future
+        Set<Consumer> consumerSet = new HashSet<>(hashRing.values().stream()
```

Comment:
can be replaced to `Set<Consumer> consumerSet = hashRing.values().stream()
            .flatMap(Collection::stream).collect(Collectors.toSet());`

Reply:
Already revised